presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java

```diff
@@ -62,22 +62,42 @@ private ExpressionTreeFilter() {}
         });
     }
 
-    public static List<Expression> extract(Function<Node, Optional<Expression>> predicate, List<? extends Node> nodes)
+    public static <T extends Expression> List<T> extractExpressionsOfTypeUsingPredicate(
+            Iterable<? extends Node> nodes,
+            Class<T> clazz,
+            Function<Node, Optional<T>> predicate)
     {
-        requireNonNull(predicate, "predicate is null");
         requireNonNull(nodes, "nodes is null");
+        requireNonNull(clazz, "clazz is null");
+        requireNonNull(predicate, "predicate is null");
 
-        ImmutableList.Builder<Expression> extractedExpressions = ImmutableList.builder();
+        ImmutableList.Builder<T> expressionsBuilder = ImmutableList.builder();
         for (Node node : nodes) {
             List<? extends Node> flattenedNodes = linearizeNodes(node);
-            extractedExpressions.addAll(flattenedNodes.stream()
+            expressionsBuilder.addAll(flattenedNodes.stream()
                     .map(predicate)
```

Comment:
I think you should first filter by class type then apply `predicate`:
```
 .filter(clazz::isInstance)
 .map(clazz:cast)
 .filter(predicate)
 .collect(toImmutableList)
```
This will allow to write simpler predicates, see: https://github.com/prestodb/presto/pull/6704/commits/f9c20b1124c096f722d8c6541df786e770682b71#r107145881